cmd/oras/root/backup.go

```diff
@@ -0,0 +1,422 @@
+		return err
+	}
+	// Create a temporary file for the tarball
+	tempTar, err := os.CreateTemp("", "oras-backup-*.tar")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary tar file: %w", err)
+	}
+	tempTarPath := tempTar.Name()
+	if err := orasio.TarDirectory(ctx, tempTar, dstRoot); err != nil {
+		return fmt.Errorf("failed to create tar archive from directory %s: %w", dstRoot, err)
+	}
+	if err := tempTar.Close(); err != nil {
+		return fmt.Errorf("failed to close temporary tar file: %w", err)
+	}
+
+	// Ensure target directory exists
+	absOutput := opts.output
+	if !filepath.IsAbs(absOutput) {
+		absOutput, err = filepath.Abs(opts.output)
+		if err != nil {
+			return fmt.Errorf("failed to get absolute path for output file %s: %w", opts.output, err)
+		}
+	}
+	if err := os.MkdirAll(filepath.Dir(absOutput), 0755); err != nil {
+		return fmt.Errorf("failed to create directory for output file %s: %w", absOutput, err)
+	}
+
+	// Move the temporary tar file to the final output path
+	if err := os.Rename(tempTarPath, absOutput); err != nil {
+		removeErr := os.Remove(tempTarPath)
+		if removeErr != nil {
+			logger.Debugf("failed to remove temporary tar file %s: %v", tempTarPath, removeErr)
+		}
+		return err
+	}
+
+	fi, err := os.Stat(absOutput)
+	if err != nil {
+		return fmt.Errorf("failed to stat output file %s: %w", absOutput, err)
+	}
+	return metadataHandler.OnTarExported(opts.output, fi.Size())
+}
+
+// resolveTags resolves tags to their descriptors.
+// It returns the resolved tags and their corresponding descriptors.
+func resolveTags(ctx context.Context, target oras.ReadOnlyTarget, specifiedTags []string) ([]string, []ocispec.Descriptor, error) {
+	var tags []string
+	var descs []ocispec.Descriptor
+	if len(specifiedTags) > 0 {
+		// resolve the specified tags
+		tags = specifiedTags
+		descs = make([]ocispec.Descriptor, 0, len(tags))
+		for _, tag := range tags {
+			desc, err := oras.Resolve(ctx, target, tag, oras.DefaultResolveOptions)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to resolve tag %q: %w", tag, err)
+			}
+			descs = append(descs, desc)
+		}
+		return tags, descs, nil
+	}
+
+	// discover all tags in the repository and resolve them
+	tagLister, ok := target.(registry.TagLister)
+	if !ok {
+		return nil, nil, errTagListNotSupported
+	}
+	if err := tagLister.Tags(ctx, "", func(gotTags []string) error {
+		for _, gotTag := range gotTags {
+			desc, err := oras.Resolve(ctx, target, gotTag, oras.DefaultResolveOptions)
+			if err != nil {
+				return fmt.Errorf("failed to resolve tag %q: %w", gotTag, err)
+			}
+			tags = append(tags, gotTag)
+			descs = append(descs, desc)
+		}
+		return nil
+	}); err != nil {
+		return nil, nil, fmt.Errorf("failed to find tags: %w", err)
+	}
+
+	return tags, descs, nil
+}
+
+// parseArtifactReferences parses the input string into a repository
+// and a slice of tags.
+func parseArtifactReferences(artifactRefs string) (repository string, tags []string, err error) {
```

Comment:
Can we reuse `registry.ParseReferenece` and work on top of it instead of manual parsing?

Think about `strings.Split(artifactRefs, ",")` first. Then do reference parsing and tag validating.

Reply:
Basically, I don't want to have parsing logics (L412-L430) here as they may be error-prone and duplicated. We should build code (e.g. reuse code) on top of `oras-go` just like we did [here](https://github.com/oras-project/oras/blob/2a3c65a67fca30cd7dfa63dbcff8508df2400c36/cmd/oras/root/push.go#L126-L127).